CvGameCoreDLL_Expansion2/CvUnit.cpp

```diff
@@ -31260,73 +31259,160 @@ int CvUnit::GetWithdrawChance(const CvUnit& attacker, const bool bCheckChances)
 
 		if (pDestPlot && !canMoveInto(*pDestPlot, MOVEFLAG_DESTINATION | MOVEFLAG_NO_EMBARK))
 		{
-			iBlockedHexes++;
+			iFreeHexes--;
 		}
 	}
 
-	// If all three hexes away from attacker blocked, we can't withdraw
-	if (iBlockedHexes >= 3)
+	return iFreeHexes;
+}
+
+int CvUnit::GetWithdrawChance(const CvUnit& attacker) const
+{
+	VALIDATE_OBJECT
+	int iWithdrawChance = getExtraWithdrawal();
+	if (iWithdrawChance == 0)
 		return 0;
-	
-	if (bCheckChances)
+
+	if (GetNumFallBackPlotsAvailable(attacker) == 0)
+		return 0;	
+
+	if (MOD_BALANCE_VP)
+	{
+		return getHasWithdrawnThisTurn() ? 0 : 100;
+	}
+	else
 	{
-		int iWithdrawChance = getExtraWithdrawal();
 		// Does attacker have a greater speed than defender? Reduce withdrawal chance for each point the attacker is faster
 		int iDefenderMovementRange = baseMoves(isEmbarked());
 		int iAttackerMovementRange = attacker.baseMoves(attacker.isEmbarked());
 		if (iAttackerMovementRange > iDefenderMovementRange)
 			iWithdrawChance += (/*-20*/ GD_INT_GET(WITHDRAW_MOD_ENEMY_MOVES) * (iAttackerMovementRange - iDefenderMovementRange));
 
-		iWithdrawChance += (/*-20*/ GD_INT_GET(WITHDRAW_MOD_BLOCKED_TILE) * iBlockedHexes);
+		iWithdrawChance += (/*-20*/ GD_INT_GET(WITHDRAW_MOD_BLOCKED_TILE) * (3 - GetNumFallBackPlotsAvailable(attacker)));
 		return iWithdrawChance;
 	}
-	else
-		return 100;
 }
 
+// returns true if the unit should withdraw from a melee attack
+bool CvUnit::CheckWithdrawal(const CvUnit& attacker) const
+{
+	VALIDATE_OBJECT
+	int iWithdrawChance = GetWithdrawChance(attacker);
+	if (iWithdrawChance == 0)
+		return false;
+
+	//include damage so the result changes for each attack
+	int iRoll = GC.getGame().getSmallFakeRandNum(10, plot()->GetPlotIndex() + GetID() + getDamage()) * 10;
+	return iRoll < iWithdrawChance;
+}
 //	--------------------------------------------------------------------------------
-bool CvUnit::DoFallBack(const CvUnit& attacker)
+bool CvUnit::DoFallBack(const CvUnit& attacker, bool bWithdraw)
 {
 	VALIDATE_OBJECT
 
 	CvPlot* pAttackerFromPlot = attacker.plot();
 	DirectionTypes eAttackDirection = directionXY(pAttackerFromPlot, plot());
 
-	int iRightOrLeftBias = (GC.getGame().getSmallFakeRandNum(10, plot()->GetPlotIndex()+GetID()+getDamage()) < 5) ? 1 : -1;
-	int iBiases[3] = {0,-1,1};
-	
-	for(int i = 0; i < 3; i++)
-	{
-		int iMovementDirection = (NUM_DIRECTION_TYPES + eAttackDirection + (iBiases[i] * iRightOrLeftBias)) % NUM_DIRECTION_TYPES;
-		CvPlot* pDestPlot = plotDirection(getX(), getY(), (DirectionTypes) iMovementDirection);
+	CvPlot* pDestPlot = NULL;
+	if (MOD_BALANCE_VP)
```

Comment:
The CP players might be confused by the fact that units would no longer retreat to the tile opposite to the attacker whenever possible.

Reply:
They would already confused by the fact that units are never able to retreat to free tiles towards the attacker as they can in the base game.

I think this improvement in functionality is worth the potential confusion; especially since it's only people who know that this is the case are those who have looked into the code themselves.